### PR TITLE
netdev: add netdev_register_signal

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR is an alternative to #16921 but more generic (as proposed by @benpicco ).

It adds a callback mechanism to indicate the upper layer that a netdev device finished its setup/registration procedure.

This function can be used as a post_setup hook to do clock configuration or trimming, without requiring to get the private descriptor through the network interface (without this PR the device descriptor is retrieved with gnrc_netif_get_by_type).

This PR is required to remove the GNRC<->netdev dependency, because it won't be guaranteed that the device descriptor of GNRC netif points to a netdev device.

### Testing procedure
Compile test should be enough.

### Issues/PRs references
Alternative to #16921

### CODE REVIEW

File: drivers/include/net/netdev.h [hunk 1]
```diff
@@ -374,6 +397,10 @@ static inline void netdev_register(struct netdev *dev, netdev_type_t type, uint8
     (void) type;
     (void) index;
 #endif
+
+    if IS_ACTIVE(CONFIG_NETDEV_REGISTER_SIGNAL_CB) {
```
[the remark this answers is not in the document]
Comment (miri64) [reply drivers/include/net/netdev.h:401]: That is not how we agreed on handling boolean defines however ;-).

[the remark this answers is not in the document]
Comment (jia200x) [reply drivers/include/net/netdev.h]: yes, otherwise Kconfig will have issues

[the remark this answers is not in the document]
Comment (jia200x) [reply drivers/include/net/netdev.h]: (anyway my header definition was wrong. I only define it now if DOXYGEN is present)

**Comment by jia200x** [comment]
I piggybacked an analogue to ac119c4185eddd5c345e4c73fa3d1dbfa20132cb

**Comment by jia200x** [comment]
may I squash?

**Comment by benpicco** [comment]
Sure, please squash

**Review by benpicco — APPROVED** [review]
still works on `openmote-b`

```
> batmon 3000
2021-10-05 00:36:15,606 # batmon 3000
2021-10-05 00:36:15,607 # NA NA NA NA NA NA NA NA NA NA NA NA NA BATMON
> batmon 2000
2021-10-05 00:36:19,686 # batmon 2000
> batmon 2600
2021-10-05 00:36:22,677 # batmon 2600
2021-10-05 00:36:22,679 # NA NA NA NA NA NA NA NA NA NA NA NA NA BATMON
2021-10-05 00:37:04,325 # 00000000  A9  2F  32  67  B9  A4  45  80  FD  DE  CE  9D  71  BF  75  44
> 2021-10-05 00:37:05,203 # get_random
2021-10-05 00:37:05,220 # 00000000  0B  CD  46  8F  6E  E4  44  F6  FA  8A  A2  07  7F  2E  92  55
> 2021-10-05 00:37:05,603 # get_random
2021-10-05 00:37:05,619 # 00000000  77  26  E0  1E  81  8C  12  68  03  BD  50  A4  6C  2C  91  3C
> 2021-10-05 00:37:05,811 # get_random
2021-10-05 00:37:05,813 # 00000000  85  4E  12  52  8C  C8  3E  B4  60  B6  A1  EE  25  0E  8D  26
> 2021-10-05 00:37:06,195 # get_random
2021-10-05 00:37:06,211 # 00000000  12  CF  A0  2C  67  F9  5C  45  C0  5E  18  CA  98  00  0A  90
> 2021-10-05 00:37:06,387 # get_random
2021-10-05 00:37:06,389 # 00000000  8E  DE  4A  96  97  9E  73  A2  37  E2  90  38  0D  49  15  95
> 2021-10-05 00:37:06,563 # get_random
2021-10-05 00:37:06,579 # 00000000  03  33  94  3E  1A  01  11  63  DC  E0  E7  38  B9  44  89  5E
> 2021-10-05 00:37:06,755 # get_random
2021-10-05 00:37:06,757 # 00000000  FA  43  97  75  EF  35  31  CE  CA  B4  E6  56  7F  0E  69  30
> 2021-10-05 00:37:06,931 # get_random
2021-10-05 00:37:06,933 # 00000000  7B  1A  8B  5E  BE  A1  00  65  DA  57  A5  C7  A2  E4  E6  17
```

```

**Comment by jia200x** [comment]
squashed!

**Comment by jia200x** [comment]
thanks for the review!

**Pull request merged**